File: app/src/main/java/com/byd/dashcast/util/concurrent/SingleFlight.java
package com.byd.dashcast.util.concurrent;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicBoolean;

/** Coalesces concurrent callers onto one leader operation and one immutable result. */
public final class SingleFlight<T> {

    private Attempt<T> current;

    public synchronized Ticket<T> join() {
        if (current == null) {
            current = new Attempt<>();
            return new Ticket<>(this, current, true);
        }
        return new Ticket<>(this, current, false);
    }

    private synchronized void clear(Attempt<T> attempt) {
        if (current == attempt) current = null;
    }

    private static final class Attempt<T> {
        private final CountDownLatch completed = new CountDownLatch(1);
        private final AtomicBoolean completionClaimed = new AtomicBoolean(false);
        private volatile T result;

        void complete(T value) {
            if (!completionClaimed.compareAndSet(false, true)) {
                throw new IllegalStateException("single-flight attempt already completed");
            }
            result = value;
            completed.countDown();
        }

        T await(long timeout, TimeUnit unit) throws InterruptedException, TimeoutException {
            if (!completed.await(timeout, unit)) {
                throw new TimeoutException("single-flight result timed out");
            }
            return result;
        }
    }

    public static final class Ticket<T> {
        private final SingleFlight<T> owner;
        private final Attempt<T> attempt;
        private final boolean leader;

        private Ticket(SingleFlight<T> owner, Attempt<T> attempt, boolean leader) {
            this.owner = owner;
            this.attempt = attempt;
            this.leader = leader;
        }

        public boolean isLeader() {
            return leader;
        }

        public void complete(T result) {
            if (!leader) throw new IllegalStateException("only the leader can complete an attempt");
            attempt.complete(result);
            owner.clear(attempt);
        }

        public T await(long timeout, TimeUnit unit) throws InterruptedException, TimeoutException {
            return attempt.await(timeout, unit);
        }
    }
}

File: app/src/main/java/com/byd/dashcast/util/concurrent/SingleFlight.kt
package com.byd.dashcast.util.concurrent

import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import java.util.concurrent.atomic.AtomicBoolean

/** Coalesces concurrent callers onto one leader operation and one immutable result. */
class SingleFlight<T> {

    private var current: Attempt<T>? = null

    @Synchronized
    fun join(): Ticket<T> {
        val existing = current
        if (existing == null) {
            val attempt = Attempt<T>()
            current = attempt
            return Ticket(this, attempt, true)
        }
        return Ticket(this, existing, false)
    }

    @Synchronized
    private fun clear(attempt: Attempt<T>) {
        if (current === attempt) current = null
    }

    internal class Attempt<T> {
        private val completed = CountDownLatch(1)
        private val completionClaimed = AtomicBoolean(false)
        @Volatile private var result: T? = null

        fun complete(value: T) {
            if (!completionClaimed.compareAndSet(false, true)) {
                throw IllegalStateException("single-flight attempt already completed")
            }
            result = value
            completed.countDown()
        }

        @Throws(InterruptedException::class, TimeoutException::class)
        fun await(timeout: Long, unit: TimeUnit): T {
            if (!completed.await(timeout, unit)) {
                throw TimeoutException("single-flight result timed out")
            }
            @Suppress("UNCHECKED_CAST")
            return result as T
        }
    }

    class Ticket<T> internal constructor(
        private val owner: SingleFlight<T>,
        private val attempt: Attempt<T>,
        private val leader: Boolean,
    ) {
        val isLeader: Boolean get() = leader

        fun complete(result: T) {
            if (!leader) throw IllegalStateException("only the leader can complete an attempt")
            attempt.complete(result)
            owner.clear(attempt)
        }

        @Throws(InterruptedException::class, TimeoutException::class)
        fun await(timeout: Long, unit: TimeUnit): T {
            return attempt.await(timeout, unit)
        }
    }
}
